core-common/src/main/java/org/zalando/nakadi/service/subscription/zk/lock/CuratorNakadiLock.java

```diff
@@ -0,0 +1,87 @@
+package org.zalando.nakadi.service.subscription.zk.lock;
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.locks.InterProcessSemaphoreMutex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.zalando.nakadi.repository.zookeeper.RotatingCuratorFramework;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Not threads safe, current usage pattern does not require it
+ */
+public class CuratorNakadiLock implements NakadiLock {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(CuratorNakadiLock.class);
+    private static final int SECONDS_TO_WAIT_FOR_LOCK = 15;
+
+    private final RotatingCuratorFramework rotatingCuratorFramework;
+    private final String lockPath;
+
+    private CuratorFramework curatorToRelease;
+    private CuratorFramework curatorLastUsed;
+    private InterProcessSemaphoreMutex semaphore;
+
+    public static String getSubscriptionLockPath(final String subscriptionId) {
+        return "/nakadi/locks/subscription_" + subscriptionId;
+    }
+
+    public CuratorNakadiLock(
+            final RotatingCuratorFramework rotatingCuratorFramework,
+            final String subscriptionId) {
+        this.rotatingCuratorFramework = rotatingCuratorFramework;
+        this.lockPath = getSubscriptionLockPath(subscriptionId);
+    }
+
+    @Override
+    public boolean lock() {
+        try {
+            curatorToRelease = rotatingCuratorFramework
+                    .takeCuratorFramework();
+            // the check is to not recreate semaphore every time
+            if (curatorToRelease != curatorLastUsed) {
+                curatorLastUsed = curatorToRelease;
+                semaphore = new InterProcessSemaphoreMutex(
+                        curatorLastUsed, lockPath);
+            }
+
+            final boolean acquired = semaphore.acquire(
+                    SECONDS_TO_WAIT_FOR_LOCK,
+                    TimeUnit.SECONDS);
+            if (!acquired) {
+                LOG.warn("failed to acquire semaphore mutex within {} seconds",
+                        SECONDS_TO_WAIT_FOR_LOCK);
+            }
+            return acquired;
+        } catch (final Exception e) {
+            LOG.error("error while acquiring semaphore mutex lock", e);
+            try {
+                rotatingCuratorFramework.returnCuratorFramework(
```

Comment:
There is a possibility that we don't have `curatorToRelease` at this point (it's `null`).

Reply:
agree

Reply:
this is fine, rotating curator know how to work with it

Reply:
i do not see how that will happen: lock and unlock methods are used by single thread only.

Reply:
It can happen if `takeCuratorFramework()` throws an exception here: https://github.com/zalando/nakadi/pull/1263/files#diff-bc88c93ce2c4b4ac1bb3d022553af0a9da93ddc3be9fb0be5c0ab54dd972453cR41